Xamarin.Essentials/SaveToGallery/SaveToGallery.android.cs

```diff
@@ -0,0 +1,113 @@
+﻿using System;
+using System.IO;
+using System.Threading.Tasks;
+using Android.Content;
+using Android.Provider;
+using Android.Webkit;
+using Environment = Android.OS.Environment;
+using File = Java.IO.File;
+using Path = System.IO.Path;
+using Stream = System.IO.Stream;
+using Uri = Android.Net.Uri;
+
+namespace Xamarin.Essentials
+{
+    public static partial class SaveToGallery
+    {
+        static async Task PlatformSaveAsync(MediaFileType type, byte[] data, string fileName, string albumName)
+        {
+            using var ms = new MemoryStream(data);
+            await PlatformSaveAsync(type, ms, fileName, albumName);
+        }
+
+        static async Task PlatformSaveAsync(MediaFileType type, string filePath, string albumName)
+        {
+            if (!System.IO.File.Exists(filePath))
+                throw new ArgumentNullException(nameof(filePath));
+            using var fileStream = System.IO.File.OpenRead(filePath);
+
+            await PlatformSaveAsync(type, fileStream, Path.GetFileName(filePath), albumName);
+        }
+
+        static async Task PlatformSaveAsync(MediaFileType type, Stream fileStream, string fileName, string albumName)
+        {
+            await Permissions.EnsureGrantedAsync<Permissions.StorageWrite>();
+
+            var context = Platform.AppContext;
+            var dateTimeNow = DateTime.Now;
+
+            var fileNameWithoutExtension = Path.GetFileNameWithoutExtension(fileName);
+            var extension = Path.GetExtension(fileName).ToLower();
+            var newFileName = $"{fileNameWithoutExtension}_{dateTimeNow:yyyyMMdd_HHmmss}{extension}";
+
+            using var values = new ContentValues();
+            values.Put(MediaStore.MediaColumns.DateAdded, TimeSeconds(dateTimeNow));
+            values.Put(MediaStore.MediaColumns.Title, fileNameWithoutExtension);
+            values.Put(MediaStore.MediaColumns.DisplayName, newFileName);
+
+            var mimeType = MimeTypeMap.Singleton.GetMimeTypeFromExtension(extension.Replace(".", string.Empty));
+            if (!string.IsNullOrWhiteSpace(mimeType))
+                values.Put(MediaStore.MediaColumns.MimeType, mimeType);
+
+            using var externalContentUri = type == MediaFileType.Image
+                ? MediaStore.Images.Media.ExternalContentUri
+                : MediaStore.Video.Media.ExternalContentUri;
+
+            var relativePath = type == MediaFileType.Image
+                ? Environment.DirectoryPictures
+                : Environment.DirectoryMovies;
+
+            if (Platform.HasApiLevelQ)
+            {
+#if MONOANDROID10_0
+                values.Put(MediaStore.MediaColumns.RelativePath, Path.Combine(relativePath, albumName));
+                values.Put(MediaStore.MediaColumns.DateTaken, TimeMillis(dateTimeNow));
```

Comment:
Would be misleading to use explicitly set the `DateTaken` to be equal to the import time (`DateAdded´).

My current scenario is downloading and importing photos taken by an external system (drone), which sets the date taken data as EXIF columns. At least this should be documented. It's fair that the Essentials should be the "base path", and in many cases, this is probably fine...

Reply:
We can try to remove this. The metadata should remain unchanged

Reply:
@follesoe, can you give me a sample file with a metadata that the drone made?

Reply:
I checked, this does not affect the date of capture a photo, which is in the metadata

|__________________________| Sample Video |__________________________|
|:---------------------------:|:-------------:|:---------------------------:|
||![Video](https://user-images.githubusercontent.com/59065470/107864216-f1f6bc80-6e6a-11eb-889a-c09c73416dd6.mov)||

Reply:
Will find a picture for you @dimonovdd, but looking at the documentation: https://developer.android.com/reference/android/provider/MediaStore.MediaColumns#DATE_TAKEN

> Indexed value of MediaMetadataRetriever#METADATA_KEY_DATE or ExifInterface#TAG_DATETIME_ORIGINAL extracted from this media item.

So I don't really see why it would make sense to set this for `DateTime.Now`, from my understanding it should not be set and then Android could index it from the EXIF metadata.

However, during our testing we did find cases where this field was not set correctly on all Android devices, so we explicitly set this field to a value manually read from the EXIF information of our file.

This is how we do it:

```csharp
var directories = ImageMetadataReader.ReadMetadata(fileStream);
var exif = directories.OfType<ExifSubIfdDirectory>().FirstOrDefault();
var dateTaken = exif?.GetDateTime(ExifDirectoryBase.TagDateTimeOriginal);

if (dateTaken.HasValue)
{
    values.Put(MediaStore.IMediaColumns.DateTaken, dateTaken.HasValue ? dateTaken.Value.ToUnixTimeMs() : dateTimeNow.ToUnixTimeMs());
}

// Set position to 0 after reading Exif.
if (fileStream.CanSeek)
{
    fileStream.Position = 0;
}
```

The EXIF is extracted using the `MetadataExtractor` NuGet package.

Reply:
Metadata remains the same. Perhaps not all devices display the date when a photo was captured which is in a metadata.
I definitely don't see the point in adding extra dependencies. 
Android documentation is good, but device manufacturers often don't follow it. That's why the documentation is useless

Reply:
@dimonovdd Sure, I am not arguing taking this dependency or reading the EXIF, but I am arguing that it makes no sense trying to set the DATE_TAKEN field at all, as the picture most definitely was NOT taken at `DateTime.Now` as the current implementation does, and Android should extract this information from the EXIF metadata so makes no sense trying to explicitly set it to a value by default.